compiler/hie-core/src/Development/IDE/Core/Compile.hs

```diff
@@ -310,29 +311,61 @@ getModSummaryFromBuffer fp contents dflags parsed = do
     , ms_parsed_mod   = Nothing
     }
 
+-- | Run CPP on a file
+runCpp :: DynFlags -> FilePath -> Maybe SB.StringBuffer -> IO SB.StringBuffer
+runCpp dflags filename contents = withTempDir $ \dir -> do
+    let out = dir </> takeFileName filename <.> "out"
+    case contents of
+        Nothing -> do
+            -- Happy case, file is not modified, so run CPP on it in-place
+            -- which also makes things like relative #include files work
+            -- and means location information is correct
+            doCpp dflags True filename out
+            liftIO $ SB.hGetStringBuffer out
+
+        Just contents -> do
+            -- Sad path, we have to create a version of the path in a temp dir
+            -- __FILE__ macro is wrong, ignoring that for now (likely not a real issue)
+
+            -- Relative includes aren't going to work, so we fix that by adding to the include path.
+            let addSelf (IncludeSpecs quote global) = IncludeSpecs (takeDirectory filename : quote) global
+            dflags <- return dflags{includePaths = addSelf $ includePaths dflags}
+
+            -- Location information is wrong, so we fix that by patching it afterwards.
+            let inp = dir </> "___HIE_CORE_MAGIC___"
+            let f x = if SB.atEnd x then Nothing else Just $ SB.nextChar x
+            liftIO $ writeFileUTF8 inp (unfoldr f contents)
```

Comment:
Converting a UTF-8 encoded `StringBuffer` to `[Char]` only to then convert it back to a UTF-8 encoded file seems needlessly wasteful. On the other hand, there also doesn’t seem to be a nice API for writing a `StringBuffer` to a file so you have to mess around with pointers and this is probably not a bottleneck either way.

Reply:
I'll add a helper function to abstract it out, then at least the silly is only in one olace

Reply:
hPutStringBuffer was added in 8.8, so I've written a compat version for 8.6 and am using that.